Authentication devise

Yuliya-Kirich opened the pull request:
I have implemented authentication in the project using the Devise library.
I have made the management of questions and answers resources accessible only to administrators.
Using the Devise helper, I have implemented the redirection of administrators to the /admin/tests page after login.
When creating a new test as an administrator, I assign the administrator as the author of the test.

Commit:
devise install

Commit:
authentication with devise

BubuntuClu commented on app/controllers/admin/base_controller.rb:
```diff
@@ -0,0 +1,14 @@
+  private
+
+  def admin_required!
+    redirect_to root_path, alert: 'Авторизуйтесь для просмотра данной страницы.' unless current_user.is_a?(Admin)
```
добавь метод в модель `User` который будет проверять является ли пользователь админом. `current_user.admin?`

BubuntuClu commented on app/controllers/admin/questions_controller.rb:
```diff
@@ -0,0 +1,64 @@
+
+  def users_spoof_check
+    if user_signed_in?
+      redirect_to root_url unless current_user.id == @question.test.user_id
```
я предпоичитаю `if ... != ...`, оно так легче понимается чем инвертирование результата

BubuntuClu commented on app/controllers/application_controller.rb:
```diff
@@ -1,26 +1,25 @@
-      redirect_to login_path, alert: 'Есть доступ в TestGuru? Проверьте пожалуйста ваш email и пароль.'
-      redirect_url_remember
+  def after_sign_in_path_for(resource)
+    if resource.is_a?(Admin)
```
`resource.admin?`

BubuntuClu commented on app/controllers/questions_controller.rb:
```diff
@@ -1,39 +1,15 @@
-  before_action :find_test, only: %i[create new]
-  before_action :find_question, only: %i[show destroy edit update]
-  before_action :users_spoof_check, only: %i[show destroy edit update]
+class Admin::QuestionsController < ApplicationController
```
это что за контроллер? для чего нужен?

Yuliya-Kirich replied:
просматривать вопросы к тесту для user

BubuntuClu commented on app/controllers/answers_controller.rb:
```diff
@@ -1,49 +1,17 @@
-class AnswersController < ApplicationController
-  before_action :find_question, only: %i[new create]
-  before_action :set_answer, only: %i[show edit update destroy]
+class Admin::AnswersController < ApplicationController
```
это что за контроллер? для чего нужен?

Yuliya-Kirich replied:
просматривать свои ответы для user

Commit:
I added a method to the User model that will check whether the user is an admin. current_user.admin?

Yuliya-Kirich commented:
I have implemented authentication in the project using the Devise library.
I have made the management of questions and answers resources accessible only to administrators.
Using the Devise helper, I have implemented the redirection of administrators to the /admin/tests page after login.
When creating a new test as an administrator, I assign the administrator as the author of the test.
